### PR TITLE
Memory management and patterns in ASP.NET Core

### LINKED ISSUE #15306: Major new doc: Memory Management and Patterns in ASP.NET Core
Memory Management and Patterns in ASP.NET Core
 
Doc will go in Performance > Memory Management and Patterns

### PR DESCRIPTION
**Pull request by Rick-Anderson**
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/memory?view=aspnetcore-3.0&branch=pr-en-us-15320)  Fixes #15306

Migration from https://github.com/sebastienros/memoryleak/blob/master/README.md

Developers who understand their topic typically write in long sentences. Long sentences don't MT (machine translate). Long sentences are difficult for native speakers to comprehend.  Verify I didn't change the meaning when splitting up long sentences. 

Please checkoff the following non-word-smithing changes:

- [ ]  Changed links from 
* old respository https://github.com/aspnet/ResponseCaching/blob/c1cb7576a0b86e32aec990c22df29c780af29ca5 
* to current https://github.com/aspnet/AspNetCore/blob/v3.0.0/src/Middleware/ResponseCaching/src/Streams/StreamUtilities.cs#L16

- [ ] Added: Is not a supported tool, however, it can be used to display memory usage patterns of ASP.NET Core apps.
- [ ]

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@Maoni0 please review.

File: aspnetcore/performance/memory.md [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+Objects are moved from one generation to another based on their lifetime. As objects live longer, they are moved into a higher generation. As mentioned previously, higher generations are GC'd less often. Short term lived objects always remain in generation 0. For example, objects that are referenced during the life of a web request are short lived. Application level singletons generally move to generation 1 and eventually 2.
+
+When an ASP.NET Core app starts, the GC:
+
+* Reserves some memory for the initial heap segments.
+* Commits a small portion of memory when the runtime is loaded.
+
+The preceding memory allocations are done for performance reasons. The performance benefit comes from heap segments in contiguous memory.
+
+**ASP.NET Core apps allocate a significant amount of memory at startup for better performance.**
```
Comment (Maoni0) [inline aspnetcore/performance/memory.md]: as a reader I'd want to know how this achieves better perf - are you caching something?

File: aspnetcore/performance/memory.md [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+
+## Analyzing the memory usage of an app
+
+Dedicated tools can help analyzing memory usage:
+
+- Counting object references
+- Measuring how much impact the GC has on CPU usage
+- Measuring memory space used for each generation
+
+Use the following tools to analyze memory usage:
```
Comment (Maoni0) [inline aspnetcore/performance/memory.md]: does VS provide something that allows you to easily run in say a production env? does it have something that doesn't require the installation of VS on the production machine? if not it'd be helpful to mention tools you could run in prod

File: aspnetcore/performance/memory.md [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+
+The following image shows the memory profile under a 5K RPS using the Workstation GC.
+
+![preceding chart](memory/_static/workstation.png)
+
+The differences between this chart and the server version are significant:
+
+- The working set drops from 500 MB to 70 MB.
+- The GC does generation 0 collections multiple times per second instead of every two seconds.
+- The GC threshold went from 300 MB to 10 MB.
```
Comment (Maoni0) [inline aspnetcore/performance/memory.md]: >The GC threshold went from 300 MB to 10 MB. [](start = 2, length = 43)

what is "GC threshold"? allocation budget? (I haven't looked at the charts)

File: aspnetcore/performance/memory.md (hunk 2 above)
Comment (davidfowl) [reply aspnetcore/performance/memory.md]: We should add dotnet trace here now

File: aspnetcore/performance/memory.md [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+* Not releasing the class correctly.
+* Forgetting to invoke the `Dispose`method of the dependent objects that should be disposed.
+
+#### Large objects heap
+
+Frequent memory allocation/free cycles can fragment memory, especially when allocating large chunks of memory. Objects are allocated in contiguous blocks of memory. To mitigate fragmentation, when the GC frees memory, it will try to defragment it. This process is called **compaction**. Compaction involves moving objects. Moving large objects imposes a performance penalty. For this reason the GC creates a special memory zone for  _large_ objects, called the [large object heap](/dotnet/standard/garbage-collection/large-object-heap) (LOH). Object that are greater than 85,000 bytes (approximately 83 KB) are:
+
+* Placed on the LOH.
+* Not compacted.
```
Comment (Maoni0) [inline aspnetcore/performance/memory.md]: >Not compacted. [](start = 2, length = 14)

LOH can be compacted at the request of the user;
in containers LOH is automatically compacted starting core 3.0

File: aspnetcore/performance/memory.md [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+* Eventually released during generation 2 collections.
```
Comment (Maoni0) [inline aspnetcore/performance/memory.md]: >Eventually  [](start = 2, length = 11)

I would say "collected during generation 2 GCs"

File: aspnetcore/performance/memory.md [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+Note: The `byte[]` structure has overhead bytes. That's why 84,976 bytes triggers the 85,000 limit.
+
+Comparing the two preceding charts:
+
+* The working set is similar for both scenarios, about 450 MB.
+* The under LOH requests (84,975 bytes) shows mostly generation 0 collections.
+* The over LOH requests generate constant generation 2 collections. Generation 2 collections are expensive. More CPU is required and throughput drops almost 50%.
+
+For maximum performance, large object use should be minimized. If possible, split up large objects. For example, [Response Caching](xref:performance/caching/response) middleware in ASP.NET Core split the cache entries into blocks less than  85,000 bytes.
```
Comment (Maoni0) [inline aspnetcore/performance/memory.md]: >For maximum performance, large object use should be minimized. If possible, split up large objects. For  [](start = 0, length = 104)

should emphasize that temporary large objects are a problem - they are the ones that cause gen2 GCs. see "Collection cost" in [https://devblogs.microsoft.com/dotnet/large-object-heap-uncovered-from-an-old-msdn-article/](https://devblogs.microsoft.com/dotnet/large-object-heap-uncovered-from-an-old-msdn-article/)

File: aspnetcore/performance/memory.md (hunk 1 above)
Comment (Rick-Anderson) [reply aspnetcore/performance/memory.md]: @sebastienros can you answer this?

Comment (davidfowl) [reply aspnetcore/performance/memory.md]: Yea, I'm a bit confused by this statement as well.

Comment (sebastienros) [reply aspnetcore/performance/memory.md]: I was probably trying to simplify the fact that the Server GC allocates some memory per core (budget). Nothing to do with ASP.NET per se. With retrospect this statement is totally wrong.

**Comment by Rick-Anderson** [comment]
@sebastienros  @davidfowl  can you approve or let me know what needs to be fixed?

File: aspnetcore/performance/memory.md [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+* The working set is constant at approximately 500 MB.
+* CPU is 33%.
+* The memory consumption and release (through GC) is stable.
+* The CPU (33%) is not over-utilized, therefore the garbage collection can keep up with a high number of allocations.
+
+### Workstation GC vs. Server GC
+
+The .NET Garbage Collector has two different modes:
+
+* **Workstation GC: Optimized for the desktop.
```
Comment (sebastienros) [inline aspnetcore/performance/memory.md]:
```suggestion
* **Workstation GC**: Optimized for the desktop.
```

File: aspnetcore/performance/memory.md [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+The preceding chart shows an obvious issue with the implementation of this class, as it keeps increasing memory usage. This is a known problem that is being tracked in [this issue](https://github.com/aspnet/Home/issues/3110).
+
+The same leak could be happen in user code, by one of the following:
+
+* Not releasing the class correctly.
+* Forgetting to invoke the `Dispose`method of the dependent objects that should be disposed.
+
+### Large objects heap
+
+Frequent memory allocation/free cycles can fragment memory, especially when allocating large chunks of memory. Objects are allocated in contiguous blocks of memory. To mitigate fragmentation, when the GC frees memory, it trys to defragment it. This process is called **compaction**. Compaction involves moving objects. Moving large objects imposes a performance penalty. For this reason the GC creates a special memory zone for _large_ objects, called the [large object heap](/dotnet/standard/garbage-collection/large-object-heap) (LOH). Object that are greater than 85,000 bytes (approximately 83 KB) are:
```
Comment (sebastienros) [inline aspnetcore/performance/memory.md]:
```suggestion
Frequent memory allocation/free cycles can fragment memory, especially when allocating large chunks of memory. Objects are allocated in contiguous blocks of memory. To mitigate fragmentation, when the GC frees memory, it trys to defragment it. This process is called **compaction**. Compaction involves moving objects. Moving large objects imposes a performance penalty. For this reason the GC creates a special memory zone for _large_ objects, called the [large object heap](/dotnet/standard/garbage-collection/large-object-heap) (LOH). Objects that are greater than 85,000 bytes (approximately 83 KB) are:
```

Comment (sebastienros) [reply aspnetcore/performance/memory.md]: The suggestion is adding an `s` on Object, in the last sentence.

**Pull request merged**